File: lib/features/club_setup/models/club_setup_draft.dart
import 'dart:typed_data';

import 'package:viro_team_v2/constants/firestore_fields.dart';
import 'package:viro_team_v2/features/club_setup/club_setup_defaults.dart';
import 'package:viro_team_v2/features/club_setup/club_setup_steps.dart';
import 'package:viro_team_v2/features/club_setup/utils/club_setup_format.dart';
import 'package:viro_team_v2/models/club.dart';
import 'package:viro_team_v2/utils/person_data_format.dart';

/// Tailles d'effectif proposées à l'onboarding (10→100 puis +25).
abstract final class ClubMemberCountRanges {
  /// Anciennes fourchettes (brouillons / analytics historiques).
  static const String under30 = 'under_30';
  static const String range30to100 = '30_100';
  static const String range100to300 = '100_300';
  static const String over300 = 'over_300';

  static final List<String> all = _buildValues();

  static List<String> _buildValues() {
    final values = <String>[];
    for (var count = 10; count <= 100; count += 10) {
      values.add('$count');
    }
    for (var count = 125; count <= 1000; count += 25) {
      values.add('$count');
    }
    return values;
  }

  static bool _isNumeric(String key) {
    if (key.isEmpty) return false;
    for (var i = 0; i < key.length; i++) {
      final code = key.codeUnitAt(i);
      if (code < 48 || code > 57) return false;
    }
    return true;
  }

  /// Libellé compact pour le sélecteur.
  static String label(String key) {
    if (_isNumeric(key)) return key;
    return switch (key) {
      under30 => '< 30',
      range30to100 => '30 – 100',
      range100to300 => '100 – 300',
      over300 => '300+',
      _ => key,
    };
  }

  /// Libellé explicite pour le récapitulatif de création.
  static String recapLabel(String key) {
    if (_isNumeric(key)) return '$key membres';
    return switch (key) {
      under30 => 'Moins de 30 membres',
      range30to100 => '30 à 100 membres',
      range100to300 => '100 à 300 membres',
      over300 => 'Plus de 300 membres',
      _ => label(key),
    };
  }

  /// Mappe une ancienne fourchette vers une valeur numérique.
  static String? migratePersisted(String? value) {
    if (value == null || value.isEmpty) return null;
    if (_isNumeric(value)) return value;
    return switch (value) {
      under30 => '20',
      range30to100 => '50',
      range100to300 => '200',
      over300 => '300',
      _ => value,
    };
  }
}

/// Brouillon local de l'assistant création de club.
class ClubSetupDraft {
  ClubSetupDraft({
    this.name = '',
    String? sport,
    this.logoBytes,
    this.logoFilePath,
    Set<String>? objectives,
    this.city = '',
    this.postalCode = '',
    this.address = '',
    List<PracticeLocation>? practiceLocations,
    this.description = '',
    this.currentStep = 0,
    this.memberCountRange,
    this.brandColorHex = ClubSetupDefaults.brandColorHex,
    this.useClubAddressAsFirstLocation = true,
  })  : sport = sport ?? ClubSports.all.first,
        objectives = objectives ?? <String>{},
        practiceLocations = practiceLocations ?? [];

  String name;
  String sport;
  Uint8List? logoBytes;
  String? logoFilePath;
  Set<String> objectives;
  String city;
  String postalCode;
  String address;
  List<PracticeLocation> practiceLocations;
  String description;
  int currentStep;
  String? memberCountRange;
  String brandColorHex;
  bool useClubAddressAsFirstLocation;

  /// Indique si le brouillon contient des données saisies (hors étape courante).
  bool get hasSavedProgress =>
      name.trim().isNotEmpty ||
      city.trim().isNotEmpty ||
      address.trim().isNotEmpty ||
      description.trim().isNotEmpty ||
      objectives.isNotEmpty ||
      practiceLocations.isNotEmpty ||
      memberCountRange != null ||
      logoFilePath != null ||
      currentStep > 0 ||
      brandColorHex != ClubSetupDefaults.brandColorHex ||
      !useClubAddressAsFirstLocation;

  bool get canProceedIdentity => name.trim().length >= 2 && sport.isNotEmpty;

  bool get canProceedObjectives => objectives.isNotEmpty;

  bool get canProceedHeadquarters {
    if (cityError(city, required: true) != null) return false;
    if (postalCodeError(postalCode) != null) return false;
    if (addressLineError(address) != null) return false;
    return true;
  }

  bool get canProceedPracticeLocations => practiceLocations.isNotEmpty;

  bool get canProceedInfo =>
      canProceedHeadquarters && canProceedPracticeLocations;

  ClubSetupDraft copy() {
    return ClubSetupDraft(
      name: name,
      sport: sport,
      logoBytes: logoBytes,
      logoFilePath: logoFilePath,
      objectives: Set<String>.from(objectives),
      city: city,
      postalCode: postalCode,
      address: address,
      practiceLocations: List<PracticeLocation>.from(practiceLocations),
      description: description,
      currentStep: currentStep,
      memberCountRange: memberCountRange,
      brandColorHex: brandColorHex,
      useClubAddressAsFirstLocation: useClubAddressAsFirstLocation,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'sport': sport,
        'logoFilePath': logoFilePath,
        'objectives': objectives.toList(),
        'city': city,
        'postalCode': postalCode,
        'address': address,
        'practiceLocations':
            practiceLocations.map((location) => location.toMap()).toList(),
        'description': description,
        'currentStep': currentStep,
        'memberCountRange': memberCountRange,
        'brandColorHex': brandColorHex,
        'useClubAddressAsFirstLocation': useClubAddressAsFirstLocation,
        'wizardVersion': ClubSetupSteps.wizardVersion,
      };

  factory ClubSetupDraft.fromJson(Map<String, dynamic> json) {
    final locationsRaw = json['practiceLocations'] as List<dynamic>? ?? [];
    final objectivesRaw = json['objectives'] as List<dynamic>? ?? [];
    final sport = json['sport'] as String? ?? ClubSports.all.first;
    final city = json['city'] as String? ?? '';
    final postalCode = json['postalCode'] as String? ?? '';
    final address = json['address'] as String? ?? '';

    final locations = ClubSetupFormat.migrateLinkedHeadquarters(
      address: address,
      postalCode: postalCode,
      city: city,
      sport: sport,
      locations: locationsRaw
          .whereType<Map<String, dynamic>>()
          .map(PracticeLocation.fromMap)
          .toList(),
    );

    final useClubAddress = json['useClubAddressAsFirstLocation'] as bool? ??
        ClubSetupFormat.linkedHeadquartersIndex(locations) >= 0;

    return ClubSetupDraft(
      name: json['name'] as String? ?? '',
      sport: sport,
      logoFilePath: json['logoFilePath'] as String?,
      objectives: objectivesRaw.whereType<String>().toSet(),
      city: city,
      postalCode: postalCode,
      address: address,
      practiceLocations: locations,
      description: json['description'] as String? ?? '',
      currentStep: ClubSetupSteps.normalizePersistedStep(
        (json['currentStep'] as num?)?.toInt() ?? 0,
        wizardVersion: (json['wizardVersion'] as num?)?.toInt() ?? 1,
      ),
      memberCountRange: ClubMemberCountRanges.migratePersisted(
        json['memberCountRange'] as String?,
      ),
      brandColorHex: json['brandColorHex'] as String? ??
          ClubSetupDefaults.brandColorHex,
      useClubAddressAsFirstLocation: useClubAddress,
    );
  }
}

File: portal/src/lib/clubSetup/clubSetupDraft.ts
import {
  CLUB_SETUP_DRAFT_KEY_PREFIX,
  CLUB_SETUP_LOGO_KEY_PREFIX,
  ClubMemberCountRanges,
  ClubSetupDefaults,
  ClubSetupSteps,
  ClubSports,
} from "./constants";
import { ClubSetupFormat } from "./clubSetupFormat";
import {
  addressLineError,
  cityError,
  postalCodeError,
} from "@/lib/format/personDataFormat";

/** Lieu de pratique du club. */
export type PracticeLocation = {
  name: string;
  /** Ville du lieu de pratique. */
  city?: string;
  address?: string;
  /** Catégorie du lieu (city-stade, forêt, lac…). */
  category?: string;
  /** Libellé libre si catégorie = Autre. */
  categoryCustom?: string;
  /** Lieu dérivé du siège (mis à jour tant que la case est cochée). */
  linkedToHeadquarters?: boolean;
};

/** Brouillon local du wizard création club. */
export type ClubSetupDraft = {
  name: string;
  sport: string;
  logoDataUrl: string | null;
  objectives: Set<string>;
  city: string;
  postalCode: string;
  address: string;
  practiceLocations: PracticeLocation[];
  description: string;
  currentStep: number;
  /** Étape la plus avancée déjà atteinte (icônes restent allumées au retour). */
  maxReachedStep: number;
  memberCountRange: string | null;
  brandColorHex: string;
  /** Case « siège = lieu de pratique » (persistée pour reprise de brouillon). */
  useClubAddressAsFirstLocation: boolean;
};

export function createEmptyClubSetupDraft(): ClubSetupDraft {
  return {
    name: "",
    sport: ClubSports.all[0],
    logoDataUrl: null,
    objectives: new Set(),
    city: "",
    postalCode: "",
    address: "",
    practiceLocations: [],
    description: "",
    currentStep: ClubSetupSteps.prerequisites,
    maxReachedStep: ClubSetupSteps.prerequisites,
    memberCountRange: null,
    brandColorHex: ClubSetupDefaults.brandColorHex,
    useClubAddressAsFirstLocation: true,
  };
}

export function canProceedIdentity(draft: ClubSetupDraft): boolean {
  return draft.name.trim().length >= 2 && draft.sport.length > 0;
}

export function canProceedObjectives(draft: ClubSetupDraft): boolean {
  return draft.objectives.size > 0;
}

export function canProceedInfo(draft: ClubSetupDraft): boolean {
  if (cityError(draft.city, { required: true })) return false;
  if (postalCodeError(draft.postalCode)) return false;
  if (addressLineError(draft.address)) return false;
  if (draft.practiceLocations.length === 0) return false;
  for (const location of draft.practiceLocations) {
    if (location.city && cityError(location.city)) return false;
    if (location.address && addressLineError(location.address)) return false;
  }
  return true;
}

export function hasSavedProgress(draft: ClubSetupDraft): boolean {
  return (
    draft.name.trim().length > 0 ||
    draft.city.trim().length > 0 ||
    draft.address.trim().length > 0 ||
    draft.description.trim().length > 0 ||
    draft.objectives.size > 0 ||
    draft.practiceLocations.length > 0 ||
    draft.memberCountRange !== null ||
    draft.logoDataUrl !== null ||
    draft.currentStep > 0 ||
    draft.brandColorHex !== ClubSetupDefaults.brandColorHex ||
    !draft.useClubAddressAsFirstLocation
  );
}

type SerializedDraft = {
  name: string;
  sport: string;
  hasLogo: boolean;
  objectives: string[];
  city: string;
  postalCode: string;
  address: string;
  practiceLocations: PracticeLocation[];
  description: string;
  currentStep: number;
  maxReachedStep?: number;
  memberCountRange: string | null;
  brandColorHex: string;
  useClubAddressAsFirstLocation?: boolean;
  wizardVersion: number;
};

export function serializeClubSetupDraft(draft: ClubSetupDraft): SerializedDraft {
  return {
    name: draft.name,
    sport: draft.sport,
    hasLogo: draft.logoDataUrl !== null,
    objectives: [...draft.objectives],
    city: draft.city,
    postalCode: draft.postalCode,
    address: draft.address,
    practiceLocations: draft.practiceLocations.map((location) => ({
      name: location.name,
      ...(location.city ? { city: location.city } : {}),
      address: location.address,
      ...(location.category ? { category: location.category } : {}),
      ...(location.categoryCustom
        ? { categoryCustom: location.categoryCustom }
        : {}),
      ...(location.linkedToHeadquarters ? { linkedToHeadquarters: true } : {}),
    })),
    description: draft.description,
    currentStep: draft.currentStep,
    maxReachedStep: draft.maxReachedStep,
    memberCountRange: draft.memberCountRange,
    brandColorHex: draft.brandColorHex,
    useClubAddressAsFirstLocation: draft.useClubAddressAsFirstLocation,
    wizardVersion: ClubSetupSteps.wizardVersion,
  };
}

export function deserializeClubSetupDraft(
  json: SerializedDraft,
  logoDataUrl: string | null,
): ClubSetupDraft {
  const wizardVersion = json.wizardVersion ?? 1;
  const currentStep = ClubSetupSteps.normalizePersistedStep(
    json.currentStep ?? 0,
    wizardVersion,
  );
  const maxReachedStep = Math.max(
    currentStep,
    json.maxReachedStep == null
      ? currentStep
      : ClubSetupSteps.normalizePersistedStep(json.maxReachedStep, wizardVersion),
  );

  const sport = json.sport ?? ClubSports.all[0];
  const city = json.city ?? "";
  const postalCode = json.postalCode ?? "";
  const address = json.address ?? "";

  const practiceLocations = ClubSetupFormat.migrateLinkedHeadquarters({
    sport,
    city,
    postalCode,
    address,
    locations: (json.practiceLocations ?? []).map((location) => ({
      name: location.name ?? "",
      ...(location.city ? { city: location.city } : {}),
      address: location.address,
      ...(location.category ? { category: location.category } : {}),
      ...(location.categoryCustom
        ? { categoryCustom: location.categoryCustom }
        : {}),
      ...(location.linkedToHeadquarters ? { linkedToHeadquarters: true } : {}),
    })),
  });

  const useClubAddressAsFirstLocation =
    typeof json.useClubAddressAsFirstLocation === "boolean"
      ? json.useClubAddressAsFirstLocation
      : ClubSetupFormat.linkedHeadquartersIndex(practiceLocations) >= 0;

  return {
    name: json.name ?? "",
    sport,
    logoDataUrl: json.hasLogo ? logoDataUrl : null,
    objectives: new Set(json.objectives ?? []),
    city,
    postalCode,
    address,
    practiceLocations,
    description: json.description ?? "",
    currentStep,
    maxReachedStep,
    memberCountRange: ClubMemberCountRanges.migratePersisted(
      json.memberCountRange,
    ),
    brandColorHex: json.brandColorHex ?? ClubSetupDefaults.brandColorHex,
    useClubAddressAsFirstLocation,
  };
}

export function draftStorageKey(userId: string): string {
  return `${CLUB_SETUP_DRAFT_KEY_PREFIX}${userId}`;
}

export function logoStorageKey(userId: string): string {
  return `${CLUB_SETUP_LOGO_KEY_PREFIX}${userId}`;
}

/** Persiste le brouillon JSON (sans le logo). */
export function persistDraftToStorage(userId: string, draft: ClubSetupDraft): void {
  if (typeof window === "undefined") return;
  const payload = serializeClubSetupDraft(draft);
  window.localStorage.setItem(draftStorageKey(userId), JSON.stringify(payload));
}

/** Charge le brouillon depuis localStorage. */
export function loadDraftFromStorage(userId: string): ClubSetupDraft | null {
  if (typeof window === "undefined") return null;
  const raw = window.localStorage.getItem(draftStorageKey(userId));
  if (!raw) return null;
  try {
    const json = JSON.parse(raw) as SerializedDraft;
    const logoDataUrl = window.localStorage.getItem(logoStorageKey(userId));
    return deserializeClubSetupDraft(json, logoDataUrl);
  } catch (error) {
    void import("@sentry/nextjs").then((Sentry) => {
      Sentry.captureException(error, {
        level: "warning",
        tags: { feature: "club_setup", area: "draft_parse" },
      });
    });
    return null;
  }
}

/** Persiste le logo en data URL. */
export function persistLogoToStorage(userId: string, logoDataUrl: string | null): void {
  if (typeof window === "undefined") return;
  const key = logoStorageKey(userId);
  if (!logoDataUrl) {
    window.localStorage.removeItem(key);
    return;
  }
  window.localStorage.setItem(key, logoDataUrl);
}

/** Efface brouillon et logo. */
export function clearDraftStorage(userId: string): void {
  if (typeof window === "undefined") return;
  window.localStorage.removeItem(draftStorageKey(userId));
  window.localStorage.removeItem(logoStorageKey(userId));
}

/** Convertit une data URL en bytes pour l’upload Storage. */
export function dataUrlToBytes(dataUrl: string): { bytes: ArrayBuffer; contentType: string } | null {
  const match = /^data:([^;]+);base64,(.+)$/.exec(dataUrl);
  if (!match) return null;
  const contentType = match[1];
  const binary = atob(match[2]);
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return { bytes: bytes.buffer, contentType };
}
